File: Java/StackQueue/Queue/CircularQueue.java
package StackQueue.Queue;

import StackQueue.Queue.Services.Queue;

public class CircularQueue implements Queue {

    private int front, rear, queueSize=5;
    private int[] circularQueue;

    public CircularQueue(int queueSize) {
        this.front = -1;
        this.rear = -1;
        if (queueSize !=0)
            this.queueSize = queueSize;
        this.circularQueue = new int[this.queueSize];
    }

    public boolean isFull(){
        if ((this.front==0 && this.rear==this.queueSize-1)||(this.front==this.rear+1))
            return true;
        return false;
    }

    public boolean isEmpty(){
        if (this.front==-1)
            return true;
        return false;
    }

    public void enqueue(int value){
        if (this.isFull()) {
            System.out.println("The circular queue is full, no room to enqueue new values");
            return;
        }
        if(this.front==-1) {
            this.front = 0;
            this.rear = 0;
            this.circularQueue[this.rear] = value;
            System.out.println("Value entered successfully");
            return;
        }
        this.rear = (this.rear+1)%this.queueSize;
        this.circularQueue[this.rear] = value;
        System.out.println("Value entered successfully");
    }

    public void dequeue(){
        if(this.isEmpty()){
            System.out.println("Queue is empty, nothing to remove");
            return;
        }
        int value = this.circularQueue[this.front];
        System.out.println("Value dequeued from the queue is: "+value);
        this.front = (this.front+1)%this.queueSize;
        if(this.front==(this.rear+1)%this.queueSize){
            this.front = -1;
            this.rear = -1;
        }
    }

    public void peek() {
        if (isEmpty()) {
            System.out.println("The Queue is Empty, nothing to peek on!!! :(");
            return;
        }
        System.out.println("The latest value entered in the queue is: "+this.circularQueue[this.rear]);
    }

    public void displayAll() {
        if (isEmpty()) {
            System.out.println("The Queue is Empty, nothing to display!!! :(");
            return;
        }
        if(this.front<=rear){
            displayLoop(this.front, this.rear);
        }else {
            displayLoop(0, this.rear);
            displayLoop(this.front, this.queueSize-1);
        }
    }

    private void displayLoop(int start, int end){
        for (int i=start; i<=end; i++){
            System.out.printf("%d : %d\n", i,this.circularQueue[i]);
        }
    }
}

File: GO-Lang/StackQueue/Queue/Services/impl/circularQueue.go
package servicesImp

import "fmt"

type CircularQueue struct {
	queueSize int
	front     int
	rear      int
	queue     []int
}

func InitCQueue(queueSize int) *CircularQueue {
	if queueSize == 0 {
		return &CircularQueue{queueSize: 5, front: -1, rear: -1, queue: make([]int, 5)}
	} else {
		return &CircularQueue{queueSize: queueSize, front: -1, rear: -1, queue: make([]int, queueSize)}
	}
}

func (cQueue *CircularQueue) IsFull() bool {
	if (cQueue.front == 0 && cQueue.rear == cQueue.queueSize-1) || cQueue.front == cQueue.rear+1 {
		return true
	}
	return false
}

func (cQueue *CircularQueue) IsEmpty() bool {
	return cQueue.front == -1
}

func (cQueue *CircularQueue) Enqueue(value int) {
	if cQueue.IsFull() {
		fmt.Println("Circular Queue is full, no room the enqueue any new value")
		return
	}

	if cQueue.front == -1 {
		cQueue.front = 0
		cQueue.rear = 0
		cQueue.queue[cQueue.rear] = value
		return
	}

	cQueue.rear = (cQueue.rear + 1) % cQueue.queueSize
	cQueue.queue[cQueue.rear] = value
}

func (cQueue *CircularQueue) Dequeue() {
	if cQueue.IsEmpty() {
		fmt.Println("Circular Queue is empty,no value to dequeue")
		return
	}

	value := cQueue.queue[cQueue.front]
	fmt.Println("Dequeued value is: ", value)

	cQueue.front = (cQueue.front + 1) % cQueue.queueSize
	if cQueue.front == (cQueue.rear+1)%cQueue.queueSize {
		cQueue.front = -1
		cQueue.rear = -1
	}
}

func (cQueue *CircularQueue) Peek() {
	if cQueue.IsEmpty() {
		fmt.Println("Circular Queue is empty,no value to peek")
		return
	}
	fmt.Println("Dequeued value is: ", cQueue.queue[cQueue.rear])
}

func (cQueue *CircularQueue) DisplayAll() {
	if cQueue.IsEmpty() {
		fmt.Println("Circular Queue is empty,no value to display")
		return
	}

	if cQueue.front <= cQueue.rear {
		displayLoop(cQueue, cQueue.front, cQueue.rear)
	} else {
		displayLoop(cQueue, 0, cQueue.rear)
		displayLoop(cQueue, cQueue.front, cQueue.queueSize-1)
	}
}

func displayLoop(cQueue *CircularQueue, start int, end int) {
	for i := start; i <= end; i++ {
		fmt.Printf("%d: %d\n", i, cQueue.queue[i])
	}
}
